server/src/main/java/org/apache/druid/metadata/input/SqlEntity.java

```
@@ -0,0 +1,198 @@
+                sql
+            ).map(
+                (index, r, ctx) -> {
+                  Map<String, Object> resultRow = foldCase ? new CaseFoldedMap() : new HashMap<>();
+                  ResultSetMetaData resultMetadata;
+                  try {
+                    resultMetadata = r.getMetaData();
+                  }
+                  catch (SQLException e) {
+                    throw new ResultSetException("Unable to obtain metadata from result set", e, ctx);
+                  }
+                  try {
+                    for (int i = 1; i <= resultMetadata.getColumnCount(); i++) {
+                      String key = resultMetadata.getColumnName(i);
+                      String alias = resultMetadata.getColumnLabel(i);
+                      Object value = r.getObject(i);
+                      resultRow.put(alias != null ? alias : key, value);
+                    }
+                  }
+                  catch (SQLException e) {
+                    throw new ResultSetException("Unable to access specific metadata from " +
+                                                 "result set metadata", e, ctx);
+                  }
+                  return resultRow;
+                }
+            ).iterator();
+            jg.writeStartArray();
+            while (resultIterator.hasNext()) {
+              jg.writeObject(resultIterator.next());
+            }
+            jg.writeEndArray();
+            jg.close();
+            return null;
+          },
+          (exception) -> {
+            final boolean isStatementException = exception instanceof StatementException ||
+                                                 (exception instanceof CallbackFailedException
+                                                  && exception.getCause() instanceof StatementException);
+            return sqlFirehoseDatabaseConnector.isTransientException(exception) && !(isStatementException);
+          }
+      );
+    }
+    return new CleanableFile()
```

The `tempFile` will not be deleted if an exception is thrown in any lines above. We should catch all exceptions and delete the file properly.